src/qililab/analog/annealing_program.py

```diff
@@ -57,37 +62,62 @@ def transpile(self, transpiler: Callable):
+        """
+        split_element = chip_element.split("_")
+        return (
+            f"{chip_element[0]}{split_element[1]}"
+            if chip_element[0] == "q"
+            else f"{chip_element[0]}{split_element[1]}_{split_element[2]}"
+        )
+
+    def get_waveforms(self, correct_xtalk: bool = True) -> dict[str, ArbitraryWave]:
         """Returns a dictionary containing (bus, waveform) for each flux control from the transpiled fluxes. `AnnealingProgram.transpile` should be run first. The waveform is an arbitrary waveform obtained from the transpiled fluxes.
 
         Returns:
-            dict[str,tuple[Bus,ArbitraryWave]]: Dictionary containing (bus, waveform) for each flux control (i.e. phix or phiz).
+            dict[str,ArbitraryWave]: Dictionary containing the waveform to be sent to each bus, with xtalk corrected
         """
-        # parse names from algorithm notation (e.g. qubit_0) to runcard notation (e.g. q0)
-        element_name_map = {"qubit": "q", "coupler": "c"}
-        circuit_element_map = {
-            (element, flux): f"{flux}_{element_name_map[element.split('_', 1)[0]]}{element.split('_', 1)[1]}"
-            for element in self._annealing_program[0].keys()
-            for flux in self._annealing_program[0][element].keys()
-            if "phi" in flux
-        }  # {(element, flux): flux_line}
-
-        # Initialize dictionary with flux_lines pointing to (corresponding bus, waveform)
-        annealing_waveforms = {  # type: ignore[var-annotated]
-            flux_line: (self._platform.get_element(flux_line), []) for flux_line in circuit_element_map.values()
-        }
-        # unravel each point of the anneal program to get timewise arrays of waveforms
-        for annealing_step in self._annealing_program:
-            for circuit_element, flux in circuit_element_map.keys():
-                annealing_waveforms[circuit_element_map[circuit_element, flux]][1].append(
-                    annealing_step[circuit_element][flux]
-                )
 
-        return {key: (value[0], ArbitraryWave(np.array(value[1]))) for key, value in annealing_waveforms.items()}
+        # Initialize maps for bus to flux and flux to bus translation
+        bus_to_flux_map = {}
+        for flux_bus in self._flux_to_bus_topology:
+            if flux_bus.flux in self._transpiled_program[0]:
+                if flux_bus.bus in bus_to_flux_map:
+                    raise ValueError(
+                        f"More than one flux pointing at bus {flux_bus.bus} in the runcard flux to bus topology"
+                    )
+                bus_to_flux_map[flux_bus.bus] = flux_bus.flux
+        flux_to_bus_map = {v: k for k, v in bus_to_flux_map.items()}
+
+        # Initialize annealing waveforms
+        annealing_waveforms = {bus: [] for bus in bus_to_flux_map}  # type: ignore[var-annotated]
+        # get xtalk matrix
+        if correct_xtalk:
+            # TODO: where do we load the CrosstalkMatrix from?
```

Comment:
Crosstalk matrix loads from calibration file. Is that what you are asking? If not please clarify what that todo really mean.

Reply:
> Is that what you are asking?

Yes. I don't know the workflow to load the calibration matrix and how to include it here. If loaded from platform, maybe we can just add the xtalk matrix from argument when we call `get_waveforms`  instead of the boolean `correct_xtalk` (or pass a `Calibration`  object, i dont know)

Reply:
passing a calibration object I think should be the way, it is mandatory for the measurement anyway at the moment, and it should remain mandatory. So when we receive an execute_annealing, we are getting the calibration file as parameter, that should help you with this doubt, no?

Reply:
Check my comment in platform to see that you already have this object loaded previously for you.